composeApp/src/commonMain/kotlin/org/ooni/probe/domain/RunDescriptors.kt

```diff
@@ -71,19 +71,32 @@ class RunDescriptors(
     private suspend fun List<Descriptor>.prepareInputs(taskOrigin: TaskOrigin) =
         map { descriptor ->
             descriptor.copy(
-                netTests = descriptor.netTests.downloadUrlsIfNeeded(taskOrigin),
-                longRunningTests = descriptor.longRunningTests.downloadUrlsIfNeeded(taskOrigin),
+                netTests = descriptor.netTests.downloadUrlsIfNeeded(taskOrigin, descriptor),
+                longRunningTests = descriptor.longRunningTests.downloadUrlsIfNeeded(taskOrigin, descriptor),
             )
         }
             .filterNot { it.allTests.isEmpty() }
 
-    private suspend fun List<NetTest>.downloadUrlsIfNeeded(taskOrigin: TaskOrigin): List<NetTest> =
-        map { test -> test.copy(inputs = test.inputsOrDownloadUrls(taskOrigin)) }
+    private suspend fun List<NetTest>.downloadUrlsIfNeeded(
+        taskOrigin: TaskOrigin,
+        descriptor: Descriptor,
+    ): List<NetTest> =
+        map { test -> test.copy(inputs = test.inputsOrDownloadUrls(taskOrigin, descriptor)) }
             .filterNot { it.test is TestType.WebConnectivity && it.inputs?.any() != true }
 
-    private suspend fun NetTest.inputsOrDownloadUrls(taskOrigin: TaskOrigin): List<String>? {
+    private suspend fun NetTest.inputsOrDownloadUrls(
+        taskOrigin: TaskOrigin,
+        descriptor: Descriptor,
+    ): List<String>? {
         if (!inputs.isNullOrEmpty() || test !is TestType.WebConnectivity) return inputs
 
+        setRunBackgroundState {
```

Comment:
This sort of assumes that the Websites descriptor will always be the first one to run. It's true for now, but we need to be careful.

Reply:
We are looping through all descriptors and tests. There's no check if it's the first one I believe.